design-proposals/nic-hotplug/nic-hotplug.md

```diff
@@ -0,0 +1,142 @@
+additional networks; if so, they would need to rely on the cluster
+administrator to provision it for them.
+
+## User Stories
+* as a user, I want to add a new network interface to a running VM.
+* as a user, I want to remove a new network interface from a running VM.
+
+## Repos
+- [CNAO](https://github.com/kubevirt/cluster-network-addons-operator)
+- [KubeVirt](https://github.com/kubevirt/kubevirt)
+- [Multus](https://github.com/k8snetworkplumbingwg/multus-cni)
+
+# Design
+
+## Multus
+Multus - a CNI plugin - only handles the ADD / REMOVE verb, and is triggered
+by kubelet only when the pod's sandbox is created - or removed. Given its
+simplicity, it assumes no networks exist whenever it is executed, and procceeds
+to call the ADD / DEL for **all** networks listed in its
+`k8s.v1.cni.cncf.io/networks` annotation.
+
+As such, the first part of the design implementation must focus on multus; it
+must be refactored to work as a thick plugin, which enables it to be triggered
+not only when the pod's sandbox is created, but also on-demand.
+
+### Thickening multus
+A “thin CNI plugin” runs as a one-shot process, typically as a binary on disk
+that runsrun on a Kubernetes host machine.
+
+A “thick CNI Plugin”, on the other hand, is a CNI component composed of two (or
+more) parts, usually composed of “shim”, and a long lived process (daemon)
+resident in memory. The “shim” is a lightweight “thin CNI plugin” component that
+simply passes CNI parameters (such as JSON configuration, and environment
+variables) to the daemon component, which then processes the CNI request.
+
+To transform multus into a thick plugin, it is needed to instantiate a long
+lived process - which will be the multus pod entrypoint - listening to a unix
+domain socket - this socket must be available both in the multus pod and the
+hosts's mount namespaces; as such, a bind mount to host this socket must be
+provided for the multus pod.
+
+The CNI shim will then be invoked by kubelet, and send the CNI ADD/DELETE
+commands to the server side of multus via the unix domain socket previously
+mentioned.
+
+Refer to the sequence diagram below for more information:
+
+![Create pod flow for thick plugin](create-pod-flow.png)
+
+### Further multus requirements
+Multus must also be updated to act only on specific networks (the ones being
+added or removed) rather than the full annotation list.
+
+Another conflicting behavior of multus is its interface naming scheme; it
+iterates all networks in the `k8s.v1.cni.cncf.io/networks` annotation, and for
+each creates an interface in the pod named `netX`, where `X=<net index> + 1`.
+The simplest way to work-around this limitation is to explicitly define the
+interface names being plugged into the pod, as documented in the
+[multus documentation](https://github.com/k8snetworkplumbingwg/multus-cni/blob/master/docs/how-to-use.md#launch-pod-with-text-annotation-with-interface-name).
```

Comment:
I think I'll add an item to the NPWG this week about the possibility of having a dynamically generated, and potentially predictable interface name, possibly as an option in Multus. Maybe it's possible that we hash a few values. I'll ask Kaloom what's done for kactus.

Reply:
Nice.

If we can create an issue out of that discussion, I'll reference it here.

Reply:
FWIW, I see in [their docs](https://github.com/kaloom/kubernetes-podagent#adding-a-network-attachment) that:

> uses consistent device naming based on the network attachment name, it’s the first 12 characters of the md5 hash of the name prefixed with net, for green it’s net9f27410725ab